### PR TITLE
feat: web3 login

### PR DESCRIPTION
**Pull request by colbr**
[JIRA](https://wilderworld.atlassian.net/browse/ZOS-293)

### What does this do?

Adds ability to sign in with web3.

Also some refactoring around `Login` to make "choose sign-in option" logic simpler, and to make it easier to add more options.

### Why are we making this change?

Need to be able to sign in with web3.

### How do I test this?

1. Open `/login` endpoint
2. Hit the `Web3 Wallet` button
3. Select your wallet of choice
4. Reject transaction, check that the error message shows up
5. Return to `Email` login, return to `Web3 Wallet` login, check error message is gone
6. Select your wallet of choice
7. Sign transaction
8. You should now be logged in

#### Things to consider:
  1. How will this affect security?
  1. How will this affect performance?
  1. Does this change any APIs?

### CODE REVIEW

File: src/store/login/saga.ts [hunk 1]
```diff
@@ -77,6 +77,14 @@ export function* web3Login(action) {
...
 }
 
+/**
+ * Switches login stage. Used for switching between login options when logging in.
+ */
+export function* switchLoginStage(action) {
+  yield put(setErrors([]));
+  yield put(setStage(action.payload));
+}
+
```
Comment (colbr) [inline src/store/login/saga.ts:87]: Wasn't sure if this was the right way to do things - I think this should probably just be held in state in `LoginContainer`, but curious to see if this was correct

Comment (dalefukami) [reply src/store/login/saga.ts:87]: > I think this should probably just be held in state in LoginContainer
No, this is the correct place. Keeping user flow logic in redux/sagas allows a greater level of control and flexibility.

There's a couple of options here, I think:
1. Add the preventative validation. But, what would you do? Just fail silently and not switch probably. Simple but doesn't really help the developer.
2. Have 2 separate functions that only do the specific thing: `switchToEmailLogin`/`switchToWeb3Login`. This has the benefit of definitely only allowing the 2 things we want. I think I prefer this option. We're only going to have 2 options so there's no concern about an explosion of login types that we have to worry about and it's very explicit what's permitted.

**Review by dalefukami — APPROVED** [review]
Looks good. I think the only comment that truly requires follow up is adding the unit test. The others are up to you. This is fine to merge and follow up with any changes.

In general I encourage questioning of conventions, etc, but they should be done in independent PRs. Deviating from the local repo conventions puts too much pressure on code reviewers and the team in general to both understand the new feature AND decide if the change of convention is beneficial all at the same time. This causes the review to be longer and applies undue pressure of, "well, I don't want to hold the feature up" when there may be disagreement on the convention change but not the feature itself, etc. My suggestion is to complete the feature while adhering to the existing conventions and then, once that PR is approved and merged, post a new PR moving to the new convention (1 convention change proposal at a time).

File: src/store/login/saga.ts (hunk 1 above)
Comment (colbr) [reply src/store/login/saga.ts:87]: For _1_ I would ideally like to set the types such that you can only provide `Email` or `Web3` as params. This only really helps for DX though, so I think _2_ is the better solution!

File: src/pages/login/login-component.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+const c = bem('login');
+
+interface LoginComponentProperties {
+  isLoggingIn: boolean;
+  onToggleLoginOption: () => void;
+  stage: LoginStage;
+}
+
+export class LoginComponent extends React.Component<LoginComponentProperties> {
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/pages/login/login-component.tsx:22]: Sounds good - I'll do them as a follow-up PR

File: src/authentication/web3-login/styles.scss [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @include flex();
+  }
+
+  &__connecting {
+    @include flex();
+
+    button {
+      text-transform: unset;
+      border: 1px solid theme.$color-greyscale-11;
+    }
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/authentication/web3-login/styles.scss:23]: I was meaning to add this variant to zUI, but would have made the scope of the issue creep even more. I'll ping you in ZERO about this one!

File: src/authentication/web3-login/container.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import { Connectors } from '../../lib/web3';
+import { RootState } from '../../store/reducer';
+import { loginByWeb3 } from '../../store/login';
+import { connectContainer } from '../../store/redux-container';
+
+import { Web3Login } from '.';
+import { Web3Connect } from '../../components/web3-connect';
+
+export interface Web3LoginContainerProperties {
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/authentication/web3-login/container.tsx:11]: Gotcha - I can do that

**Pull request merged**